File: packages/agent_core/src/agent_core/domain/schemas/session.py
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field


class ExplanationPayload(BaseModel):
    type: Literal["explanation"] = "explanation"
    definition: str = Field(min_length=1)
    core_principles: list[str] = Field(min_length=1)
    worked_example: str = Field(min_length=1)
    common_mistake: str = Field(min_length=1)
    next_step: str = Field(min_length=1)


class HintPayload(BaseModel):
    type: Literal["hint"] = "hint"
    hint_level: Literal["conceptual", "scaffolded", "targeted"]
    next_step_hint: str = Field(min_length=1)
    key_principle: str = Field(min_length=1)
    pitfall: str = Field(min_length=1)
    encouragement: str = Field(min_length=1)
    direct_answer_given: bool = False


AssistantPayload = ExplanationPayload | HintPayload


class MessageTurnMetrics(BaseModel):
    history_count: int
    memory_context_count: int
    cross_session_context_count: int
    hint_level: Literal["conceptual", "scaffolded", "targeted"] | None = None
    hint_history_count: int = 0
    used_quiz_context: bool = False
    used_error_analysis: bool = False
    retrieval_latency_ms: int
    llm_latency_ms: int
    llm_retry_count: int
    response_shape_valid: bool


class CreateSessionRequest(BaseModel):
    learner_profile_id: str | None = Field(default=None, max_length=36)
    learner_goal_id: str | None = Field(default=None, max_length=36)
    title: str | None = Field(default=None, max_length=255)
    subject: str | None = Field(default=None, max_length=255)


class SessionResponse(BaseModel):
    id: str
    learner_profile_id: str
    learner_goal_id: str | None = None
    daily_task_id: str | None = None
    title: str | None
    subject: str | None
    status: str
    message_count: int
    last_activity_at: datetime
    summary: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class UpdateSessionStatusRequest(BaseModel):
    status: str = Field(min_length=1, max_length=32)


class MessageRequest(BaseModel):
    content: str = Field(min_length=1, max_length=4000)
    mode: Literal["chat", "hint"] = "chat"
    related_quiz_id: str | None = Field(default=None, max_length=36)
    question_prompt: str | None = Field(default=None, max_length=4000)
    learner_answer: str | None = Field(default=None, max_length=4000)


class MessageResponse(BaseModel):
    session_id: str
    user_message_id: str
    assistant_message_id: str
    assistant_message: str
    assistant_payload: AssistantPayload
    skill_trace: list[str]
    turn_metrics: MessageTurnMetrics


class MessageHistoryItemResponse(BaseModel):
    id: str
    session_id: str
    role: str
    content: str
    mode: str | None
    skill_trace: list[str]
    content_payload: AssistantPayload | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MessageHistoryResponse(BaseModel):
    items: list[MessageHistoryItemResponse]
    total: int
    next_before_id: str | None

File: packages/frontend/src/types/session.ts
export interface Session {
  id: string;
  learner_profile_id: string;
  learner_goal_id: string | null;
  daily_task_id: string | null;
  title: string | null;
  subject: string | null;
  status: string;
  message_count: number;
  last_activity_at: string;
  summary: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateSessionRequest {
  learner_profile_id?: string;
  learner_goal_id?: string;
  title?: string;
  subject?: string;
}

export interface UpdateSessionStatusRequest {
  status: string;
}

export interface ExplanationPayload {
  type: "explanation";
  definition: string;
  core_principles: string[];
  worked_example: string;
  common_mistake: string;
  next_step: string;
}

export interface HintPayload {
  type: "hint";
  hint_level: "conceptual" | "scaffolded" | "targeted";
  next_step_hint: string;
  key_principle: string;
  pitfall: string;
  encouragement: string;
  direct_answer_given: boolean;
}

export type AssistantPayload = ExplanationPayload | HintPayload;

export interface MessageTurnMetrics {
  history_count: number;
  memory_context_count: number;
  cross_session_context_count: number;
  hint_level: "conceptual" | "scaffolded" | "targeted" | null;
  hint_history_count: number;
  used_quiz_context: boolean;
  used_error_analysis: boolean;
  retrieval_latency_ms: number;
  llm_latency_ms: number;
  llm_retry_count: number;
  response_shape_valid: boolean;
}

export interface MessageResponse {
  session_id: string;
  user_message_id: string;
  assistant_message_id: string;
  assistant_message: string;
  assistant_payload: AssistantPayload;
  skill_trace: string[];
  turn_metrics: MessageTurnMetrics;
}

export interface MessageRequest {
  content: string;
  mode?: "chat" | "hint";
  related_quiz_id?: string;
  question_prompt?: string;
  learner_answer?: string;
}

export interface MessageHistoryItem {
  id: string;
  session_id: string;
  role: string;
  content: string;
  mode: string | null;
  skill_trace: string[];
  content_payload: AssistantPayload | null;
  created_at: string;
}

export interface MessageHistoryResponse {
  items: MessageHistoryItem[];
  total: number;
  next_before_id: string | null;
}
